### PR TITLE
refactor: Improve communication between plugin and ui & bundle ui 

### PR DESCRIPTION
**Pull request by agerardin**
The plugin is relying on a timer that triggers a full refresh every second.
This impacts UI responsiveness and increase resource usage.
The UI script also uses a timer to figure out when the d3 library has finished loading.

This commit addresses both issues by:

* Replacing the first timer by a poll mechanism. When the API notifies
  on a change, a model change is added to a queue. The ui checks the queue
  for new updates and apply them. This allows for a cleaner execution model
  and improves responsiveness.

* Using extra-scripts to bundle ui & d3 with webpack. This removes the
  second timer and have the side benefit of reducing the plugin's size by
  5 due to treeshaking.

### CODE REVIEW

**Comment by treymo** [comment]
Thank you for this @agerardin!! I'll review this today or tomorrow and get it merged

**Comment by treymo** [comment]
I'm testing this out locally and seeing: `Module not found: Error: Can't resolve 'd3' in '$HOME/dev/joplin-graph-view/src/ui'`

I think the d3.js file should remain checked into the repo.  Are you installing that dependency somewhere outside the repository on your machine?

**Review by treymo — CHANGES_REQUESTED** [review]
I'm testing this out locally and seeing: `Module not found: Error: Can't resolve 'd3' in '$HOME/dev/joplin-graph-view/src/ui'`

I think the d3.js file should remain checked into the repo. Are you installing that dependency somewhere outside the repository on your machine?

**Comment by agerardin** [comment]
Hello! I have added d3 as a dependency in ```package.json```.  ```ui/index.js``` is declared as an extraScript in ```plugin.config.json``` so it will be bundled by webpack in a separate entry. rerun ```npm install``` should get rid of the error.

File: src/index.ts [hunk 1]
```diff
@@ -105,16 +108,51 @@ async function fetchData() {
   return data;
 }
 
+
+async function redraw() {
```
Comment (treymo) [inline src/index.ts:112]: Is this function used, if not, please remove

File: src/index.ts [hunk 2]
```diff
@@ -49,8 +48,12 @@ async function getFilteredNotes(notes: Map<string, joplinData.Note>,
   return filteredNotes;
 }
 
+var count = 0;
+
 async function fetchData() {
   const selectedNote = await joplin.workspace.selectedNote();
+  const selectedFolder = await joplin.workspace.selectedFolder();
```
Comment (treymo) [inline src/index.ts:55]: Is this used?

**Review by treymo — CHANGES_REQUESTED** [review]
Just a couple minor questions, but this is a great improvement! Thank you for bringing some real JS experience to the project haha

**Comment by agerardin** [comment]
Hello! I have updated with your remarks. 
The promise callback mechanism is still quite hacky but that's the best I could come out with using the current webviewAPI. I am delving into Joplin's plugin code to see how to do better  but it will take some more time.

**Review by treymo — APPROVED** [review]

**Comment by agerardin** [comment]
To follow up on this, I was looking for a mechanism to send message from the plugin to the view rather than polling the plugin for changes. This is currently not implemented. I have a pull request ready for enabling push notifications so I will start discussing it with Joplin's contributors. Until this happens, the current method will do!

**Pull request merged**